open-craft/configuration
# [BB-216] Fix syncing logs when shutting down a server

This fixes syncing logs when rebooting, powering off or halting the instance.
Ubuntu [adopted](https://en.wikipedia.org/wiki/Upstart#Adoption) systemd in 15.04 version and therefore my proposition is to use systemd service directive for sending logs before shutdown.

I decided to run the syncing script on exit of `syslog.service`, because it allows syncing logs before rsyslogd is terminated.

I've also extracted commands for syncing logs from `playbooks/roles/vhost/templates/etc/init/sync-on-stop.conf.j2` to `playbooks/roles/vhost/templates/sync-logs-on-exit.j2` to avoid code repetition.

**Sandbox IP**: 54.38.41.167

**Note**: I'm using the `agrendalath/hawthorn.1-s3_exit-sync` config, because this one has S3 syncing fixed, which allows to test the solution.
**Note2**: This PR should not be merged before #53. The redeployment should be done after merging both of them to make it fully working.

**Testing instructions:**
ACTIONS = {`REBOOT`, `POWER OFF`, `HALT`}
For each ACTION in $ACTIONS:
  - Create an instance with `config_version = 'agrendalath/hawthorn.1-s3_exit-sync'` (or use an existing one).
  - Add something to logs: e.g. `echo TEST | sudo tee --append /edx/var/log/tracking/tracking.log`.
  - $ACTION the server and 
  - List the logs: `s3cmd --access_key $AWS_S3_LOGS_ACCESS_KEY_ID --secret_key $AWS_S3_LOGS_SECRET_KEY ls s3://$COMMON_OBJECT_STORE_LOG_SYNC_BUCKET/$COMMON_OBJECT_STORE_LOG_SYNC_PREFIX$(ec2metadata --security-groups | head -1)/$(ec2metadata --instance-id)-$(ec2metadata --local-ipv4)/` (you can get the keys from appserver 'Configuration' in Ocim or from the `/edx/bin/send-logs-to-object-store` file).
  - Get the log file (it should be the last one): `s3cmd --access_key $AWS_S3_LOGS_ACCESS_KEY_ID --secret_key $AWS_S3_LOGS_SECRET_KEY get s3://$COMMON_OBJECT_STORE_LOG_SYNC_BUCKET/$COMMON_OBJECT_STORE_LOG_SYNC_PREFIX$(ec2metadata --security-groups | head -1)/$(ec2metadata --instance-id)-$(ec2metadata --local-ipv4)/$FILE_NAME`.
  - Ensure that the downloaded file contains the sent phrase: e.g. with: `zgrep TEST $DOWNLOADED_FILE`.
